### PR TITLE
avm1: Set __constructor__ whenever we construct an object

### PR DESCRIPTION
**Pull request by Dinnerbone**
I don't think this is exhaustive and we may still be forgetting places. The method in which we create objects may need to be rethought soon.

### CODE REVIEW

**Comment by Herschel** [comment]
Thanks! I'm going to hold off on this until #939 is in, because I think this will build on top of that (Can we wire up the constructor property in `construct`?)

TODO ideas for cleaning up our object model further:

 * We shouldn't need to keep both "constructor" and "prototype" references in SystemPrototypes -- if we have the constructor object, we can grab constructor.proto whenever we need it.

 *  Eventually remove `TObject::new` in favor of #939 `construct`. `TObject::new` only serves to "make some native TObject type" -- but this should be job of the construction Function itself (see #701) and not called on the prototype. Naming-wise, `construct` also follows the ES terminology of `[[Construct]]`.

**Comment by Herschel** [comment]
Ok, I think all of the wiring can be done inside `Function::construct` now, and this removes the need for `TObject::set_constructor` as well.

**Comment by Dinnerbone** [comment]
Looking into it now.

I'm not sure we can hook it up in `Function::construct` as that may be too late (if the constructor does something with the newly constructed object, it may already require the `__constructor__` to be set). The order should always be: make new object, assign `__proto__` and `__constructor__`, do something. That means it needs to be in whatever is allocating the object, to be safest.

**Comment by Herschel** [comment]
Yea, you should be able to do it at the top of `construct` before the actual "constructor function" runs. I agree that it needs to be cleaned up, but that's a little bit of a bigger change (#701). My thought is as follows:

 * `new` is merged into `construct`, so `construct` is our single call for "create this object".
 * `construct` does not take `this`, because it creates `this` and should return it, and does the following:
    * Wire up constructor/prototype properties.
    * Run `Function::constructor`.
       * For most types that don't have a special constructor, create a ScriptObject `this` and run the normal `call`. (This can done by making the default `constructor` that creates the object and forwards that object along to `call`).
       * Primitive types run their own special `constructor` as is done currently.

**Comment by Dinnerbone** [comment]
I've done that. I was tempted to already merge `new` and `construct` however I realized that there's one special case that will complicate it: Movies have a `StageObject` which will wrap the provided thing, so they need their object before they construct it. I don't know how to neatly solve that.

**Comment by Herschel** [comment]
Could StageObjects still manually fire their constructor with `call`, and not go thru `construct`? That seems okay to me, given their weirdo construction process, it's going to be a special case in some way.

That does mean maybe I was hasty to suggest removing `set_constructor`, as a StageObject would still need to use this to rig up their `__constructor__` in `post_instantiation`. 

Another idea is to add a special `StageObject::stage_construct` method that does the above for stage instances (rigs up properties and manually calls the constructor function without creating a this).

FWIW, I did confirm that when manually wiring up a prototype for a MovieClip subclass (a la #710), you can totally omit calling `super` and all of the properties like `_x` still work (unlike Sound and others, which require you to call `super`):

```
function Bar() {
	// super(); not necessary
	trace(this._x); // works
	this._x = 200;  // works
	this.stop(); // works only if prototype was wired up
}

Bar.prototype = new Object();
Bar.prototype.__proto__ = MovieClip.prototype;
Bar.prototype.__constructor__ = MovieClip;

Object.registerClass("Bar", Bar);
attachMovie("Bar", "bar", 0);

var bar = new Bar(); // no good in AS2; properties and methods return undefined.
```

Since `super()` doesn't really do anything and `new MovieClip()` is a no-go in AS2, maybe `StageObject::construct` is a no-op/returns a dummy ScriptObject.

**Comment by Dinnerbone** [comment]
> Could StageObjects still manually fire their constructor with `call`, and not go thru `construct`? That seems okay to me, given their weirdo construction process, it's going to be a special case in some way.
> 
> That does mean maybe I was hasty to suggest removing `set_constructor`, as a StageObject would still need to use this to rig up their `__constructor__` in `post_instantiation`.

If the object construction was done by `Function::construct` and not the actual constructor, then that works. However...


> Another idea is to add a special `StageObject::stage_construct` method that does the above for stage instances (rigs up properties and manually calls the constructor function without creating a this).

I thought about that too but perhaps the actual construction of an object should probably be done by the actual constructor itself? The Sound constructor should be the only one who cares that it makes Sound objects, otherwise you're going to have to say `array.construct(..., sound_proto_to_make_new_from)`. If that's true, that then would mean two actual constructor functions which is just nasty. I feel that if you want to construct an object you should only need to know what the constructor is, and no other special requirements for that specific constructor.

-----

Thinking about this some more (I edited this comment way too many times), there's a middleground I didn't consider.

We pass in the prototype of the object-to-be-made in `FunctionObject::constructor()`. It's optional right now but nothing provides `None`, so let's change that not to be optional. At the beginning of a theoretical `Object::construct_new()` method, it creates an object by something like `self.get("prototype").new()` and then it calls `Object::construct_on_existing()` with that new value, where that method is basically what we have today (with setting `__constructor__` and calling a method).

Movieclips go with `construct_on_existing`, everything else changes from `let object = some_proto.new(); some_constructor.construct(object)` to `let object = some_constructor.construct_new()`.

Does that sound good?

**Comment by Herschel** [comment]
Sounds good to me; that's what I had in mind in the above suggestions -- the actual creation of the object should be done by the `constructor`.

Your `construct_on_existing` is the same as my `StageObject::stage_construct`, we're just going by different names. Let's go with `construct` and `construct_on_existing`.

**Comment by Dinnerbone** [comment]
There's a lot of places where we're still making objects "manually", via methods like `ScriptObject::object`. Those should be reviewed and ideally changed over, but, I think this PR is already overscoping.

The latest commits make it easy to construct objects correctly, and fixes up everything that is already calling the constructor.

File: core/src/avm1/globals/array.rs [hunk 1]
```diff
@@ -136,10 +136,8 @@ pub fn array_function<'gc>(
 ) -> Result<Value<'gc>, Error<'gc>> {
     let mut consumed = false;
 
-    let p = activation.avm.prototypes.array;
-    let array_obj = p.new(activation, context, p, &[])?;
     let constructor = activation.avm.prototypes.array_constructor;
-    constructor.construct(activation, context, array_obj, args)?;
+    let array_obj = constructor.construct(activation, context, args)?;
```
Comment (Herschel) [inline core/src/avm1/globals/array.rs:140]: This shouldn't set `__constructor__`

File: core/src/avm1/globals/context_menu.rs [hunk 2]
```diff
@@ -20,10 +20,8 @@ pub fn constructor<'gc>(
 
     this.set("onSelect", callback.into(), activation, context)?;
 
-    let obj_proto = activation.avm.prototypes.object;
-    let built_in_items = obj_proto.new(activation, context, obj_proto, &[])?;
     let constructor = activation.avm.prototypes.object_constructor;
-    constructor.construct(activation, context, built_in_items, &[])?;
+    let built_in_items = constructor.construct(activation, context, &[])?;
```
Comment (Herschel) [inline core/src/avm1/globals/context_menu.rs:24]: Should not set `__constructor__`

File: core/src/avm1/globals/shared_object.rs [hunk 3]
```diff
@@ -114,18 +114,17 @@ fn recursive_deserialize<'gc>(
...
-                let _ = constructor.construct(activation, context, obj, &[]);
-                recursive_deserialize(JsonValue::Object(o.clone()), activation, obj, context);
-
-                object.define_value(
-                    context.gc_context,
-                    entry.0,
-                    Value::Object(obj),
-                    EnumSet::empty(),
-                );
+                if let Ok(obj) = constructor.construct(activation, context, &[]) {
```
Comment (Herschel) [inline core/src/avm1/globals/shared_object.rs:118]: Should not set `__constructor__`

File: core/src/avm1/globals/shared_object.rs [hunk 4]
```diff
@@ -158,20 +157,16 @@ pub fn get_local<'gc>(
...
     // Set the internal name
     let obj_so = this.as_shared_object().unwrap();
     obj_so.set_name(action_context.gc_context, name.to_string());
 
     // Create the data object
-    let data_proto = activation.avm.prototypes.object;
-    let data = data_proto.new(activation, action_context, so, &[])?;
     let constructor = activation.avm.prototypes.object_constructor;
-    constructor.construct(activation, action_context, data, &[])?;
+    let data = constructor.construct(activation, action_context, &[])?;
```
Comment (Herschel) [inline core/src/avm1/globals/shared_object.rs:169]: Should not set `__constructor__

**Comment by Herschel** [comment]
We have to be a little careful because many of these things don't actually wire up the constructor properties, only things that use `new` (or built-ins that would do so). Inline array/object initialization does not, so for example:

```
var a = Array(2);
trace(a.hasOwnProperty("__constructor__")); // Currently true, should be false
trace(a.__constructor__);
trace("");

var a = new Array(2);
trace(a.hasOwnProperty("__constructor__")); // true
trace(a.__constructor__); // function
trace("");
```

**Comment by Dinnerbone** [comment]
Darnit Flash, you don't make things easy 😅

**Comment by Dinnerbone** [comment]
Undone those examples and rebased with latest master!

**Comment by Herschel** [comment]
Thank you!

**Pull request merged**